### PR TITLE
Beautify company

### PR DESCRIPTION
**Pull request by tuhdo**
Make company looks more harmonious with themes in general by using the
supported theme faces.

Solarized light demo:

beautify-company-solarized-light

Zenburn demo:

beautify-company-zenburn

Jazz demo:

beautify-company-jazz

### CODE REVIEW

**Comment by CestDiego** [comment]
Can you do something about the popup too? :) also I don't think the completions' background should be the same color as the default background...maybe try darkening it out a bit?

**Comment by tuhdo** [comment]
Yea I will try to on that. Btw, here is how it looks in monokai:

beautify-company-monokai

**Comment by tuhdo** [comment]
Another screenshot from monokai with candidate annotation:

beautify-company-monokai2

**Comment by CestDiego** [comment]
omg that feels so natural. We can darken the background color with [hexrgb](http://emacswiki.org/emacs/hexrgb.el) package

**Comment by tuhdo** [comment]
hmm why darken the background though?

**Comment by CestDiego** [comment]
It get's totally lost in with the buffer background....

**Comment by tuhdo** [comment]
Well that's on purpose.

**Comment by CestDiego** [comment]
not even a border? :p

**Comment by tuhdo** [comment]
A border is made with background of each candidate. You cannot have a border without changing the background color of each candidate.

File: contrib/auto-completion/packages.el [hunk 1]
```diff
@@ -98,11 +98,21 @@
...
+                          :background (face-attribute 'highlight :background))
+      ;; just disable the below face, it's redundant
+      (set-face-attribute 'company-tooltip-common-selection nil
+                          :background nil)
+      (set-face-attribute 'company-tooltip-annotation nil
+                          :foreground (face-attribute 'font-lock-constant-face :foreground))
+      (set-face-attribute 'company-tooltip nil
+                          :foreground (face-attribute 'default :foreground)
+                          :background (face-attribute 'default :background))
+      (def)
```
Comment (CestDiego) [inline contrib/auto-completion/packages.el]: lolwat

Comment (tuhdo) [inline contrib/auto-completion/packages.el]: Fixed.

**Comment by CestDiego** [comment]
I understand now :(

**Comment by trishume** [comment]
I dislike the fact that there is no difference in the background colour for the popup. It looks confusing and it means that you rely on reading text to tell what is a candidate and what isn't because the scrollbar isn't always there.

:-1:  from me. I much prefer the way the company popup looks right now. That might be because I designed it to be exactly how I wanted it when I wrote the original company layer.

**Comment by syl20bnr** [comment]
It lacks a drop shadow :-)
Seriously, I find this confusing too, it is hard to differentiate the menu from the buffer.

**Comment by CestDiego** [comment]
it seriously lacks a drop shadow @syl20bnr ! that would make everything easier. I experimented with using :box and :underline attributes but it's not the same.....I will give a more thorough try later.

**Comment by tonylotts** [comment]
If the completion background is lighter than the default background, that might help it stand out.

**Comment by trishume** [comment]
I thought of a solution that should work, we can just automatically use a subtly contrasting colour for the popup background, chosen based on the theme! I took the liberty of going back in time and implementing this 4 months ago, It's currently in master :stuck_out_tongue_winking_eye: 

I seriously don't think the current company popup needs any improvement. A drop shadow would make it better but since emacs doesn't do drop shadows I think it is fine.

**Comment by CestDiego** [comment]
@trishume are you referring to the current implementaion? :O I don't see it being a contrasting colour. D:

**Comment by trishume** [comment]
@CestDiego I said *subtly* contrasting. IMO it shouldn't be a significant contrast such that the colours have to be inverted, but it should at least have a subtle difference.

In solarized light by default it looks like this:
screen shot 2015-05-05 at 10 09 14 pm

**Comment by tuhdo** [comment]
The background color could be adjusted. I made this PR because the current company does not look nice outside of some supported theme and the default yellow theme is a bit ugly.

**Comment by syl20bnr** [comment]
I find the current colors for company consistent with all the theme I use:
- solarized light/dark
- leuven
- zenburn
- gotham
- material

capture d ecran 2015-05-05 a 22 44 35
capture d ecran 2015-05-05 a 22 44 58
capture d ecran 2015-05-05 a 22 45 18
capture d ecran 2015-05-05 a 22 45 30
capture d ecran 2015-05-05 a 22 46 18
capture d ecran 2015-05-05 a 22 46 54

@tuhdo Do you have a screenshot of a bad looking one ?

I don't know if we can do better than the current colors, they feel natural enough for me to never wonder if we should change them.

**Comment by syl20bnr** [comment]
Side note: @trishume we can see on the screenshot my issue with `ESC` returning in normal state without closing company. I have to investigate this.

**Comment by tuhdo** [comment]
@CestDiego It's the themes from megapack and elsewhere. Here I am using jazz theme and it uses default company faces.

**Comment by syl20bnr** [comment]
So the color tweak of @trishume is not enabled in jazz, I don't have time to look into this tonight but maybe we can do something about it.

**Comment by tuhdo** [comment]
@syl20bnr @CestDiego @trishume I updated it. Now the background color of company popup has slightly darker color compared with the default background color, just enough to make it look nice. You can adjust the `-0.04` value to your liking i.e. lighter or darker.

- Solarized light:

beautify-company-solarized-light

- Solarized dark:

beautify-company-solarized-dark

- Zenburn:

beautify-company-zenburn

- Leuven:
beautify-company-leuven

- Jazz:
beautify-company-jazz

- Gotham:

beautify-company-gotham

**Comment by CestDiego** [comment]
luv it!! :D

**Comment by tuhdo** [comment]
Let's compare the old and new company in solarized light:

- New:

beautify-company-solarized-light

- Old:

so-so-company-solarized-light

The improvements:

- Scroll bar is actually visible, so we know the position of current candidate.
- The highlighter is subtle and does not change text properties of current candidate. Imagine `hl-line` face that changes text properties of the line it highlights: annoying at best, unusable at worst.
- Uses less colors, so it's less distracting.
- Background color of the popup is subtle and adaptable. It's slightly more visible than the default background color, but just enough to blend with the surrounding harmoniously.
- Use proper method to set faces.

**Comment by trishume** [comment]
I definitely like some of these changes:
- I like the non-bold face making spacing proper
- I like the normal colour for candidates
- Proper method of setting faces

I'm on the fence over these, I'm leaning towards not liking them but could be convinced otherwise:
- Less obvious highlight. I don't like how dramatic the current highlight is but I also don't like the idea of using text colour to indicate highlight. It would better fit with UI conventions and IMO look better if it was just a more subtle background colour change (but not as subtle as the almost-invisible change in your screenshot).
- Scroll bar track. I think it looks much cleaner and integrated without the track. With a decent background contrast it doesn't add any value in terms of seeing where the bar is.
- I prefer the existing common prefix highlighting. I think the grayscale-bold fits better. The colours make me think "keyword" and don't flow as well IMO. To use your own words the bold "uses less colors, so it's less distracting"

I don't like:
- Uniformly subtle darkened background. The method you use to darken the background looks decent in some themes and awful in others. The current method of asking the theme leads to better integration. I don't like the idea of users having to customize the company popup darkening for their theme when we already have an automatic solution that is decent. Theme authors have already chosen a nice alternative background colour, lets take their advice.

**Comment by tuhdo** [comment]
> Less obvious highlight. I don't like how dramatic the current highlight is but I also don't like the idea of using text colour to indicate highlight. It would better fit with UI conventions and IMO look better if it was just a more subtle background colour change (but not as subtle as the almost-invisible change in your screenshot).

Here is a Helm example:

- Highlighter with color and change text properties, like the current company (the theme is wombat):

helm-wombat

- Subtle highlighter that actually just highlight and nothing else (the theme is Zenburn):

helm-zenburn

Probably the above screenshot on light themes (like Solarized light) are a bit hard to see, but if you really try this PR in your Emacs, it's actually visible enough without intrusively change text colors, similar to the Zenburn screenshot above.

> Scroll bar track. I think it looks much cleaner and integrated without the track. With a decent background contrast it doesn't add any value in terms of seeing where the bar is.

Hmm probably. It may look better if we also remove the gray box at the end of current candidate.

> I prefer the existing common prefix highlighting. I think the grayscale-bold fits better. The colours make me think "keyword" and don't flow as well IMO. To use your own words the bold "uses less colors, so it's less distracting"

It looks not so nice on dark theme. But I can use `font-lock-comment-face`, since that face is usually gray and is designed to fit with current theme. We also don't use company in comments (most of the time) anyway. So it stands out quite well.

> Uniformly subtle darkened background. The method you use to darken the background looks decent in some themes and awful in others. The current method of asking the theme leads to better integration. I don't like the idea of users having to customize the company popup darkening for their theme when we already have an automatic solution that is decent. Theme authors have already chosen a nice alternative background colour, lets take their advice.

Yea probably the background color should be left for the theme designer. However, we should have a check: If two themes have the same default company colors, so we better apply this setting.

**Comment by syl20bnr** [comment]
I like the visible scroll bar. This is the only thing I want to change from the current colors. 

While being nice @tuhdo's version is not as consistent as the current colors. For instance in dark themes the company background is not visible which is a no go. It happens at a lesser degree for the highlighted candidate.

Here what I suggest:
- import @tuhdo's scroll bar setting which is better
- find a way to make jazz (and other themes) look good with the current settings, could be a recipe to follow.

**Comment by tuhdo** [comment]
@syl20bnr Are you sure the background in dark themes is not visible? It is quite visible for Solarized Dark:
beautify-company-solarized-dark


- Or Jazz:

beautify-company-jazz

Please click on the images to view at full size. Github resized it smaller to fit on a page, so the colors are not clear. And you can always make it darker with the function `spacemacs/rgb-add`.

**Comment by tuhdo** [comment]
With my current commit, jazz or other themes is properly set if it is the default. It doesn't work when switching themes with `SPC T h` though, because the face setup code only runs when company is init. We will need to add hook to run such code to refresh faces for packages when switching themes.

**Comment by syl20bnr** [comment]
In `gotham` and `jazz` I don't see them. `gotham` not at all how hard I focus, `jazz` I can see it when I focus but I should not have to do this, it should be obvious. In `solarized-dark` it is ok.

**Comment by syl20bnr** [comment]
Actually I can see it in `gotham` but when I see it, it disappear a second later. Trust me, this is painful. It may be my screen but I don't want to play with my screen settings for a theme ;-)

**Comment by tuhdo** [comment]
Ok so how about the highlighter? Do you like how company highlighter looks like Helm in `wombat` or like Helm in Zenburn?

**Comment by syl20bnr** [comment]
For the change theme issue there is a function in spacemacs that is already hooked: https://github.com/syl20bnr/spacemacs/blob/develop/core/core-themes-support.el#L205

**Comment by syl20bnr** [comment]
I prefer the zenburn highlighter because it fits the theme and my preference. But not all theme should have subtle colors. Each theme has its personality. In all the screenshots I posted, none of the highlighters were ugly to me, to be honest I only noticed the scroll bar.
The wombat is ugly but it is part of the theme, I'm sure somebody will like it and then use wombat as his main theme. There are so many themes out there that we should not bother too much about the colors except when the issue comes from a package (and not the theme).

**Comment by CestDiego** [comment]
I like the jazz theme I think I'm going to try it out..btw is there any conclussion on this?

**Comment by trishume** [comment]
Personally I say close or modify. IMO these changes are a step backwards for everything except the jazz theme. There are some good things in here but currently the bad outweighs the good. See my previous comment for what I like and don't like.

I saw your jazz theme company screenshot and that is indeed super ugly. I recommend you submit a PR upstream to the theme, or insert a special case for it in your dotfiles or spacemacs.

**Comment by syl20bnr** [comment]
Here is what I suggest for this PR:
- keep @trishume config and add @tuhdo scrollbar config
- find a way to support jazz with the current config or submit PR upstream to fix this ugly yellow background that destroy the theme.

**Comment by tuhdo** [comment]
I will change according to @syl20bnr comment but later.

**Comment by syl20bnr** [comment]
Closing this PR for inactivity. Feel free to re-open a new one against current develop if you want.

**Pull request closed without merging**